From: Jani Nikula <jani.nikula@intel.com>
Date: Wed Oct 21 17:22:43 2015 +0300
Subject: drm/i915: add hotplug activation period to hotplug update mask
Patch-mainline: Queued in subsystem maintainer repository
Git-commit: f9e3dc786a585e1ec371c446c62dcc77d0548629
Git-repo: git://anongit.freedesktop.org/drm-intel
References: bsc#953980

commit 0706f17c307b056ff6f1848320ba82d76945a6ff
Author: Egbert Eich <eich@suse.de>
Date:   Wed Sep 23 16:15:27 2015 +0200

    drm/i915: Avoid race of intel_crt_detect_hotplug() with HPD interrupt, v2

added a check with WARN to ensure only bits within the mask are
enabled. Turns out that doesn't hold for G4X, which spits out:

[    2.641439] ------------[ cut here ]------------
[    2.641444] WARNING: CPU: 0 PID: 1 at drivers/gpu/drm/i915/i915_irq.c:182 i915_hotplug_interrupt_update_locked+0x45/0x83()
[    2.641446] WARN_ON(bits & ~mask)
etc.

Add CRT_HOTPLUG_ACTIVATION_PERIOD_64 to the mask to fix the warning.

Reported-and-tested-by: Oleksij Rempel <linux@rempel-privat.de>
References: https://bugzilla.kernel.org/show_bug.cgi?id=104991
Fixes: 0706f17c307b ("drm/i915: Avoid race of intel_crt_detect_hotplug() with HPD interrupt, v2")
Cc: Egbert Eich <eich@suse.de>
Cc: Daniel Vetter <daniel.vetter@ffwll.ch>
Signed-off-by: Jani Nikula <jani.nikula@intel.com>
Reviewed-by: Daniel Vetter <daniel.vetter@ffwll.ch>
Link: http://patchwork.freedesktop.org/patch/msgid/1445437363-3030-1-git-send-email-jani.nikula@intel.com
Signed-off-by: Egbert Eich <eich@suse.com>
---
 drivers/gpu/drm/i915/i915_irq.c |    6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -2937,8 +2937,10 @@ static void i915_hpd_irq_setup(struct dr
 		hotplug_en |= CRT_HOTPLUG_VOLTAGE_COMPARE_50;
 
 		/* Ignore TV since it's buggy */
-		i915_hotplug_interrupt_update_locked(dev_priv, HOTPLUG_INT_EN_MASK
-					             | CRT_HOTPLUG_VOLTAGE_COMPARE_MASK,
+		i915_hotplug_interrupt_update_locked(dev_priv,
+						     HOTPLUG_INT_EN_MASK |
+						     CRT_HOTPLUG_VOLTAGE_COMPARE_MASK |
+						     CRT_HOTPLUG_ACTIVATION_PERIOD_64,
 					             hotplug_en);
 	}
 }
